File: src/main/java/xyz/chunkstories/core/converter/mappings/Sign.java
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.converter.mappings;

import xyz.chunkstories.api.converter.mappings.NonTrivialMapper;
import xyz.chunkstories.api.exceptions.world.WorldException;
import xyz.chunkstories.api.voxel.Voxel;
import xyz.chunkstories.api.voxel.components.VoxelComponent;
import xyz.chunkstories.api.world.World;
import xyz.chunkstories.api.world.cell.FutureCell;
import xyz.chunkstories.api.world.chunk.Chunk;
import xyz.chunkstories.core.voxel.VoxelSign;
import xyz.chunkstories.core.voxel.components.VoxelComponentSignText;
import io.xol.enklume.MinecraftChunk;
import io.xol.enklume.MinecraftRegion;
import io.xol.enklume.nbt.*;
import io.xol.enklume.util.SignParseUtil;

public class Sign extends NonTrivialMapper {

	public Sign(Voxel voxel) {
		super(voxel);
	}

	@Override
	public void output(World csWorld, int csX, int csY, int csZ, int minecraftBlockId, int minecraftMetaData,
			MinecraftRegion region, int minecraftCuurrentChunkXinsideRegion, int minecraftCuurrentChunkZinsideRegion,
			int x, int y, int z) {

		Chunk chunk = csWorld.getChunksManager().getChunkWorldCoordinates(csX, csY, csZ);
		assert chunk != null;

		if (voxel instanceof VoxelSign) {

			if (!voxel.getName().endsWith("_post")) {
				if (minecraftMetaData == 2)
					minecraftMetaData = 8;
				else if (minecraftMetaData == 3)
					minecraftMetaData = 0;
				else if (minecraftMetaData == 4)
					minecraftMetaData = 4;
				else if (minecraftMetaData == 5)
					minecraftMetaData = 12;
			}

			FutureCell future = new FutureCell(csWorld, csX, csY, csZ, voxel);
			future.setMetaData(minecraftMetaData);
			csWorld.pokeSimple(future);

			try {
				translateSignText(csWorld.tryPeek(csX, csY, csZ).getComponents().getVoxelComponent("signData"),
						region.getChunk(minecraftCuurrentChunkXinsideRegion, minecraftCuurrentChunkZinsideRegion), x, y,
						z);
			} catch (WorldException e) {
				e.printStackTrace();
			}
			// TODO Move Sign text getting here ?

		} else
			System.out.println("fuck you 666");
	}

	private void translateSignText(VoxelComponent target, MinecraftChunk minecraftChunk, int x, int y, int z) {
		NBTCompound root = minecraftChunk.getRootTag();

		if (root == null)
			return;

		NBTList entitiesList = (NBTList) root.getTag("Level.TileEntities");

		// Check it exists and is of the right kind
		if (target instanceof VoxelComponentSignText) {
			VoxelComponentSignText signTextComponent = (VoxelComponentSignText) target;
			signTextComponent.setSignText("<corresponding sign not found :(>");

			for (NBTag element : entitiesList.elements) {
				NBTCompound entity = (NBTCompound) element;
				NBTString entityId = (NBTString) entity.getTag("id");

				int tileX = ((NBTInt) entity.getTag("x")).data;
				int tileY = ((NBTInt) entity.getTag("y")).data;
				int tileZ = ((NBTInt) entity.getTag("z")).data;

				if (entityId.data.toLowerCase().equals("sign")
						|| entityId.data.toLowerCase().equals("minecraft:sign")) {
					if ((tileX & 0xF) != x || tileY != y || (tileZ & 0xF) != z) {
						continue;
					}

					String text1 = SignParseUtil.parseSignData(((NBTString) entity.getTag("Text1")).data);
					String text2 = SignParseUtil.parseSignData(((NBTString) entity.getTag("Text2")).data);
					String text3 = SignParseUtil.parseSignData(((NBTString) entity.getTag("Text3")).data);
					String text4 = SignParseUtil.parseSignData(((NBTString) entity.getTag("Text4")).data);

					String textComplete = text1 + "\n" + text2 + "\n" + text3 + "\n" + text4;
					signTextComponent.setSignText(textComplete);
					break;

				}
			}
		}
	}
}

File: src/main/java/xyz/chunkstories/core/converter/mappings/Sign.kt
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.converter.mappings

import io.xol.enklume.MinecraftChunk
import io.xol.enklume.MinecraftRegion
import io.xol.enklume.nbt.NBTCompound
import io.xol.enklume.nbt.NBTInt
import io.xol.enklume.nbt.NBTList
import io.xol.enklume.nbt.NBTString
import io.xol.enklume.util.SignParseUtil
import xyz.chunkstories.api.block.BlockType
import xyz.chunkstories.api.converter.NonTrivialMapper
import xyz.chunkstories.api.world.World
import xyz.chunkstories.api.world.cell.PodCellData
import xyz.chunkstories.api.world.chunk.ChunkCell
import xyz.chunkstories.core.voxel.VoxelSign
import xyz.chunkstories.core.voxel.components.SignData

class Sign(blockType: BlockType) : NonTrivialMapper(blockType) {

    override fun output(csWorld: World, csX: Int, csY: Int, csZ: Int, minecraftBlockId: Int, minecraftMetaData: Int,
                        region: MinecraftRegion, minecraftCuurrentChunkXinsideRegion: Int, minecraftCuurrentChunkZinsideRegion: Int,
                        x: Int, y: Int, z: Int) {
        var minecraftMetaData = minecraftMetaData

        if (blockType is VoxelSign) {

            if (!blockType.name.endsWith("_post")) {
                when (minecraftMetaData) {
                    2 -> minecraftMetaData = 8
                    3 -> minecraftMetaData = 0
                    4 -> minecraftMetaData = 4
                    5 -> minecraftMetaData = 12
                }
            }

            csWorld.setCellData(csX, csY, csZ, PodCellData(blockType))

            (csWorld.getCellMut(csX, csY, csZ) as ChunkCell).additionalData["signData"]?.let {
                translateSignText(it as SignData, region.getChunk(minecraftCuurrentChunkXinsideRegion, minecraftCuurrentChunkZinsideRegion), x, y, z)
            }
        } else {
            throw Exception("Inconsistent behavior, expected sign")
        }
    }

    private fun translateSignText(target: SignData, minecraftChunk: MinecraftChunk, x: Int, y: Int, z: Int) {
        val root = minecraftChunk.rootTag ?: return

        val entitiesList = root.getTag("Level.TileEntities") as NBTList

        target.signText = "<corresponding sign not found :(>"

        for (element in entitiesList.elements) {
            val entity = element as NBTCompound
            val entityId = entity.getTag("id") as NBTString

            val tileX = (entity.getTag("x") as NBTInt).data
            val tileY = (entity.getTag("y") as NBTInt).data
            val tileZ = (entity.getTag("z") as NBTInt).data

            if (entityId.data.toLowerCase() == "sign" || entityId.data.toLowerCase() == "minecraft:sign") {
                if (tileX and 0xF != x || tileY != y || tileZ and 0xF != z) {
                    continue
                }

                val text1 = SignParseUtil.parseSignData((entity.getTag("Text1") as NBTString).data)
                val text2 = SignParseUtil.parseSignData((entity.getTag("Text2") as NBTString).data)
                val text3 = SignParseUtil.parseSignData((entity.getTag("Text3") as NBTString).data)
                val text4 = SignParseUtil.parseSignData((entity.getTag("Text4") as NBTString).data)

                val textComplete = text1 + "\n" + text2 + "\n" + text3 + "\n" + text4
                target.signText = textComplete
                break

            }
        }
    }
}